apps/ios/SimpleXChat/ChatTypes.swift

```diff
@@ -3352,14 +3366,43 @@ public enum CIGroupInvitationStatus: String, Decodable {
 
 public enum RcvDirectEvent: Decodable {
     case contactDeleted
+    case profileUpdated(fromProfile: Profile, toProfile: Profile)
 
     var text: String {
         switch self {
         case .contactDeleted: return NSLocalizedString("deleted contact", comment: "rcv direct event chat item")
+        case let .profileUpdated(fromProfile, toProfile): return profileUpdatedText(fromProfile, toProfile)
+        }
+    }
+
+    private func profileUpdatedText(_ from: Profile, _ to: Profile) -> String {
+        return if to.displayName != from.displayName || to.fullName != from.fullName {
+            String.localizedStringWithFormat(NSLocalizedString("contact %@ changed to %@", comment: "rcv direct event chat item"), from.profileViewName, to.profileViewName)
+        } else if to.image != from.image {
+            if to.image == nil {
+                profileUpdatedRemovedPictureText
+            } else {
+                profileUpdatedSetNewPictureText
+            }
+        } else if to.contactLink != from.contactLink {
+            if to.contactLink == nil {
+                NSLocalizedString("removed contact address", comment: "rcv direct event chat item")
+            } else {
+                NSLocalizedString("set new contact address", comment: "rcv direct event chat item")
+            }
+        } else {
+            // shouldn't happen if backend correctly creates item; UI should be synchronized with backend
+            defaultProfileUpdatedText
         }
     }
 }
 
+private var profileUpdatedRemovedPictureText: String = NSLocalizedString("removed profile picture", comment: "rcv direct or group event chat item")
+
+private var profileUpdatedSetNewPictureText: String = NSLocalizedString("set new profile picture", comment: "rcv direct or group event chat item")
+
+private var defaultProfileUpdatedText: String = NSLocalizedString("updated profile", comment: "rcv direct or group event chat item")
```

Comment:
just define them in-place, they are used twice, and NSLocalizedString is probably not needed there?